### ISSUE TITLE
What does the cosine_scale mean?

### ISSUE THREAD

**Issue by Jianghao-Li** [ExplainableML/czsl#30]
Why multiply a self.scale in front of the predicted probability? Because before CE loss the probability will pass a softmax, i think the scale is meaningless, why is the result so much worse when I set the scale equal to 1?

**Comment by mancinimassimiliano** [comment]
Hi @Jianghao-Li and sorry for the late reply! 

You are right that there is a softmax in between. However, the cosine similarity is bounded between -1 and 1, making the resulting probability "flat" and causing a variety of issues (mostly related to convergence). The scale fixes this problem, making the probability span the full range [0,1].

As a simple example, if you have classes A and B with scores `[-1,1]`, if `scale=1`, you get probabilities after softmax of `[0.1192, 0.8808]`, while with `scale = 10` you get `[0,1]`.

To find more technical (and precise) explanations, you may check:
 [Zhang et. al, "AdaCos: Adaptively Scaling Cosine Logits for Effectively Learning Deep Face Representations", CVPR 2019.](https://openaccess.thecvf.com/content_CVPR_2019/papers/Zhang_AdaCos_Adaptively_Scaling_Cosine_Logits_for_Effectively_Learning_Deep_Face_CVPR_2019_paper.pdf)
